swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift

```
@@ -38,89 +38,87 @@ class PacketTunnelProvider: NEPacketTunnelProvider {
+      // connected. The system will try to restart us with a fresh config
+      // once the user fixes the problem, but we'd rather not connect
+      // without a slug.
+      completionHandler(
+        PacketTunnelProviderError.savedProtocolConfigurationIsInvalid(
+          "providerConfiguration.accountSlug"
         )
-        return
-      }
-
-      Telemetry.accountSlug = accountSlug
-
-      let internetResourceEnabled: Bool = if let internetResourceEnabledJSON = providerConfiguration[VPNProfileManagerKeys.internetResourceEnabled]?.data(using: .utf8) {
-        (try? JSONDecoder().decode(Bool.self, from: internetResourceEnabledJSON )) ?? false
-      } else {
-        false
-      }
-
-      let adapter = Adapter(
-        apiURL: apiURL,
-        token: token,
-        id: id,
-        logFilter: logFilter,
-        internetResourceEnabled: internetResourceEnabled,
-        packetTunnelProvider: self
       )
-      self.adapter = adapter
+      return
+    }
 
-      do { try await adapter.start() }
-      catch {
-        Log.error(error)
-        completionHandler(error)
+    Telemetry.accountSlug = accountSlug
 
-        return
-      }
+    let internetResourceEnabled: Bool = if let internetResourceEnabledJSON = providerConfiguration[VPNProfileManagerKeys.internetResourceEnabled]?.data(using: .utf8) {
+      (try? JSONDecoder().decode(Bool.self, from: internetResourceEnabledJSON )) ?? false
+    } else {
+      false
+    }
 
-      // Tell the system the tunnel is up, moving the tunnel manager status to
-      // `connected`.
-      completionHandler(nil)
+    let adapter = Adapter(
+      apiURL: apiURL,
+      token: token,
+      id: id,
+      logFilter: logFilter,
+      internetResourceEnabled: internetResourceEnabled,
+      packetTunnelProvider: self
+    )
+    self.adapter = adapter
+
+    do { try adapter.start() }
+    catch {
+      Log.error(error)
+      completionHandler(error)
```

Do we define the completion handler? If yes, wouldn't it be better to log there?

---

That would also allow us to learn about the other cases where we early-exit from this function due to an error.

---

Unfortunately not - this comes from the overridden startTunnel function from PacketTunnelProvider.

---

We only exit from this function due to error if the adapter start call fails - this is the main change of the PR this one is stacked on.

In other words, only the adapter.start call can block connecting the tunnel; for all others we simply log to sentry and continue.

---

> Unfortunately not - this comes from the overridden startTunnel function from PacketTunnelProvider.

But we could easily wrap it right? Like saying:

```rust
let completionHandler = |e| {
   Log.error(e);
   originalCompletionHandler(e);
};
```

---

Ah, yes I see what you mean. We don't want to log all errors though. We need to return an error in `completionHandler` to signal to the OS the tunnel was not started, but we don't necessarily need to log when the token is not available for example. That will happen if the user toggles on the VPN in system settings.